doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
 Match over terms
 ~~~~~~~~~~~~~~~~
 
-Ltac2 features a construction similar to Ltac1 :n:`match` over terms, although
+Ltac2 features a construction similar to Ltac1 :tacn:`match` over terms, although
 in a less hard-wired way.
 
-.. productionlist:: coq
-   ltac2_term       : match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-                    : lazy_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-                    : multi_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-   constrmatching  : | `constrpattern` => `ltac2_term`
-   constrpattern   : `term`
-                   : context  [ `term` ]
-                   : context `lident` [ `term` ]
-
-This construction is not primitive and is desugared at parsing time into
-calls to term matching functions from the `Pattern` module. Internally, it is
-implemented thanks to a specific scope accepting the :n:`@constrmatching` syntax.
-
-Variables from the :n:`@constrpattern` are statically bound in the body of the branch, to
+.. tacn:: @ltac2_match_key @ltac2_expr__term with @ltac2_match_list end
+   :name: lazy_match!; match!; multi_match!
+
+   .. insertprodn ltac2_match_key ltac2_match_pattern
+
+   .. prodn::
+      ltac2_match_key ::= lazy_match!
+      | match!
+      | multi_match!
+      ltac2_match_list ::= {? %| } {+| @ltac2_match_rule }
+      ltac2_match_rule ::= @ltac2_match_pattern => @ltac2_expr
+      ltac2_match_pattern ::= @cpattern
+      | context {? @ident } [ @cpattern ]
+
+   Evaluates :n:`@ltac2_expr__term`, which must yield a term, and matches it
+   sequentially with the :token:`ltac2_match_pattern`\s, which may have
+   metavariables.  When a match is found, metavariable values are substituted
+   into :n:`@ltac2_expr`, which is then applied.
+
+   Matching may continue depending on whether  `lazy_match!`, `match!` or `multi_match!`
+   is specified.
+
+   .. todo PR I see ? is sometimes need in the ltac2_expr and sometimes not, but
+      I don't when which is required
+
+   NOT RIGHT YET: In the :token:`ltac2_match_pattern`\s, metavariables have the form :n:`?@ident`, whereas
+   in the :n:`@ltac2_expr`\s, the question mark is omitted.  Choose your metavariable
+   names with care to avoid name conflicts.  For example, if you use the metavariable `S`,
+   then the :token:`ltac2_expr` can't use `S` to refer to the constructor of `nat`
+   without qualifying the constructor as `Datatypes.S`.
```

Comment:
Ltac2 variable names are not allowed to be capital letters.  So if your metavariable starts with a capital letter, you cannot use it at all.  Additionally, unlike in Ltac1, I believe conflicts are *not* an issue, because you have to antiquote Ltac2 variables.  So if you name your metavariable `?nat`, there is still a difference between `constr:(nat)` and `constr:($nat)` (or whatever the correct anti-quotation is; I can't quite recall offhand).

But this is already covered below with
```
Note that unlike Ltac, only lowercase identifiers are valid as Ltac2
bindings, so that there will be a syntax error if one of the bound variables
starts with an uppercase character.
```

Reply:
I dropped the 3 sentences starting with "Choose ...".  The first sentence is the same question as above about when a '?' is required on a variable name.